### PR TITLE
[ZEPPELIN-4273] Support Flink 1.9 for Flink Interpreter

### PR DESCRIPTION
**Pull request by zjffdu**
### What is this PR for?
This PR is to support latest flink 1.9.0.  Main features:
* Support table api
* Support batch sql
* Support streaming sql
* Support both flink planner and blink planner.
* Support pyflink

### What type of PR is it?
[ Improvement | Feature | Documentation ]

### Todos
* [ ] - Task

### What is the Jira issue?
* https://jira.apache.org/jira/browse/ZEPPELIN-4273

### Questions:
* Does the licenses files need update? No
* Is there breaking changes for older versions? NO
* Does this needs documentation? No

### CODE REVIEW

File: flink/pom.xml [hunk 1]
```diff
@@ -37,11 +37,14 @@
   <properties>
     <!--library versions-->
     <interpreter.name>flink</interpreter.name>
-    <flink.version>1.7.1</flink.version>
-    <flink.akka.version>2.3.7</flink.akka.version>
+    <flink.version>1.9-SNAPSHOT</flink.version>
```
Comment (felixcheung) [inline flink/pom.xml]: snapshot?

File: flink/pom.xml [hunk 2]
```diff
@@ -123,52 +209,287 @@
...
+          <groupId>org.apache.hive</groupId>
+          <artifactId>hive-metastore</artifactId>
+        </exclusion>
+      </exclusions>
+    </dependency>
+
+    <dependency>
+      <groupId>org.apache.flink</groupId>
+      <artifactId>flink-table-planner_2.11</artifactId>
+      <version>${flink.version}</version>
```
Comment (felixcheung) [inline flink/pom.xml]: why is this not provided?

File: flink/src/main/java/org/apache/zeppelin/flink/FlinkSqlInterrpeter.java [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+      }
+      sqlCommands.add(sqlCommand.get());
+    }
+    for (SqlCommandParser.SqlCommandCall sqlCommand : sqlCommands) {
+      try {
+        callCommand(sqlCommand, context);
+        context.out.flush();
+      }  catch (Throwable e) {
+        LOGGER.error("Fail to run sql:" + sqlCommand.operands[0] + "\n"
+                + ExceptionUtils.getStackTrace(e));
```
Comment (felixcheung) [inline flink/src/main/java/org/apache/zeppelin/flink/FlinkSqlInterrpeter.java]: why not `LOGGER.error("Fail to run sql:" + sqlCommand.operands[0], e);`?

File: flink/src/main/java/org/apache/zeppelin/flink/sql/RetractStreamSqlJob.java [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+      context.out.write("%table\n");
+      for (int i = 0; i < schema.getFieldCount(); ++i) {
+        String field = schema.getFieldName(i).get();
+        context.out.write(field);
+        if (i != (schema.getFieldCount() - 1)) {
+          context.out.write("\t");
+        }
+      }
+      context.out.write("\n");
+      LOGGER.debug("*****************Row size: " + materializedTable.size());
```
Comment (felixcheung) [inline flink/src/main/java/org/apache/zeppelin/flink/sql/RetractStreamSqlJob.java]: remove?

File: flink/src/main/java/org/apache/zeppelin/flink/sql/RetractStreamSqlJob.java [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+          }
+        }
+        LOGGER.debug("Row:" + row);
+        context.out.write("\n");
+      }
+
+      context.out.write("\n%text ");
+      context.out.flush();
+    } catch (IOException e) {
+      e.printStackTrace();
```
Comment (felixcheung) [inline flink/src/main/java/org/apache/zeppelin/flink/sql/RetractStreamSqlJob.java]: remove printStackTrace

File: flink/src/main/java/org/apache/zeppelin/flink/sql/TimeSeriesStreamSqlJob.java [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+        long maxTimestamp =
+                ((java.sql.Timestamp) materializedTable.get(materializedTable.size() - 1)
+                .getField(0)).getTime();
+
+        materializedTable = materializedTable.stream()
+                .filter(row -> ((java.sql.Timestamp) row.getField(0)).getTime() >
+                        maxTimestamp - tsWindowThreshold)
+                .collect(Collectors.toList());
+
+        LOGGER.debug("*****************Row size: " + materializedTable.size());
```
Comment (felixcheung) [inline flink/src/main/java/org/apache/zeppelin/flink/sql/TimeSeriesStreamSqlJob.java]: remove?

File: flink/src/main/java/org/apache/zeppelin/flink/sql/TimeSeriesStreamSqlJob.java [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+              context.out.write("\t");
+            }
+          }
+          LOGGER.debug("Row:" + row);
+          context.out.write("\n");
+        }
+      }
+      context.out.flush();
+    } catch (IOException e) {
+      e.printStackTrace();
```
Comment (felixcheung) [inline flink/src/main/java/org/apache/zeppelin/flink/sql/TimeSeriesStreamSqlJob.java]: remove?

there seems to be some duplicated code here?

Comment (zjffdu) [reply flink/src/main/java/org/apache/zeppelin/flink/sql/TimeSeriesStreamSqlJob.java]: fixed

File: flink/src/main/java/org/apache/zeppelin/flink/sql/TimeSeriesStreamSqlJob.java (hunk 6 above)
Comment (zjffdu) [reply flink/src/main/java/org/apache/zeppelin/flink/sql/TimeSeriesStreamSqlJob.java]: fixed

File: flink/src/main/java/org/apache/zeppelin/flink/sql/RetractStreamSqlJob.java (hunk 5 above)
Comment (zjffdu) [reply flink/src/main/java/org/apache/zeppelin/flink/sql/RetractStreamSqlJob.java]: fixed

File: flink/src/main/java/org/apache/zeppelin/flink/sql/RetractStreamSqlJob.java (hunk 4 above)
Comment (zjffdu) [reply flink/src/main/java/org/apache/zeppelin/flink/sql/RetractStreamSqlJob.java]: fixed

File: flink/src/main/java/org/apache/zeppelin/flink/FlinkSqlInterrpeter.java (hunk 3 above)
Comment (zjffdu) [reply flink/src/main/java/org/apache/zeppelin/flink/FlinkSqlInterrpeter.java]: fixed

File: flink/pom.xml (hunk 2 above)
Comment (zjffdu) [reply flink/pom.xml]: fixed

**Comment by zjffdu** [comment]
Any more comments ? @felixcheung

**Comment by f7753** [comment]
Looking forward to this change. Any progress?

File: flink/pom.xml (hunk 1 above)
Comment (zjffdu) [reply flink/pom.xml]: fixed

**Comment by f7753** [comment]
I just clone the master branch and build a tar, then deploy it on a node to connect to a exesting flink 1.9 cluster running on kubernetes. It just showed that it can not connect to the remote interpreter.

**Comment by zjffdu** [comment]
@f7753 Please ask question in user mail list and also attach necessary context, such as how you configure the interpreter and the interpreter log.

**Comment by f7753** [comment]
@zjffdu OK got it. But seems no reply. I' ll try to send another mail.

**Pull request closed without merging**